components/org.wso2.carbon.identity.application.authenticator.kakao/src/main/java/org/wso2/carbon/identity/application/authenticator/kakao/KakaoCustomAuthenticator.java

```diff
@@ -76,21 +123,251 @@ public List<Property> getConfigurationProperties() {
+            requiredParameterList.add(ACCESS_TOKEN_PARAM);
+            requiredParameterList.add(ID_TOKEN_PARAM);
+            authenticatorData.setPromptType(FrameworkConstants.AuthenticatorPromptType.INTERNAL_PROMPT);
+            authenticatorData.setAdditionalData(getAdditionalData(context, true));
+        } else {
+            requiredParameterList.add(OAUTH2_GRANT_TYPE_CODE);
+            requiredParameterList.add(OAUTH2_PARAM_STATE);
+            authenticatorData.setPromptType(FrameworkConstants.AuthenticatorPromptType.REDIRECTION_PROMPT);
+            authenticatorData.setAdditionalData(getAdditionalData(context, false));
+        }
+        authenticatorData.setRequiredParams(requiredParameterList);
+
+        return Optional.of(authenticatorData);
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
+                                                 AuthenticationContext context) throws AuthenticationFailedException {
+
+        super.initiateAuthenticationRequest(request, response, context);
+        String clientId = context.getAuthenticatorProperties().get(KakaoCustomAuthenticatorConstants.CLIENT_ID);
+        String state = context.getContextIdentifier() + KAKAO_OAUTH2_STATE_SUFFIX;
+        String redirectUri = context.getAuthenticatorProperties().get(CALLBACK_URL);
+        context.setProperty(STATE, state);
+        StringBuilder redirectUrl = new StringBuilder(KAKAO_AUTH_URL)
+                .append("?response_type=code")
+                .append("&client_id=").append(clientId)
+                .append("&redirect_uri=").append(redirectUri)
+                .append("&state=").append(state);
+        context.setProperty(REDIRECT_URL, redirectUrl.toString());
+    }
+
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
+                                                 AuthenticationContext context) throws AuthenticationFailedException {
+
+        try {
+            Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+            String token;
+            if (isNativeSDKBasedFederationCall(request)) {
+                token = request.getParameter(ACCESS_TOKEN_PARAM);
+                String idToken = request.getParameter(ID_TOKEN_PARAM);
+                if (StringUtils.isNotBlank(idToken)) {
+                    try {
+                        validateJWTToken(context, idToken);
+                    } catch (Exception e) {
+                        setAuthenticatorMessageToContext(
+                                KakaoCustomAuthenticatorConstants.ErrorMessages.JWT_TOKEN_VALIDATION_FAILED,
+                                context);
+                        throw new AuthenticationFailedException(KakaoCustomAuthenticatorConstants.ErrorMessages.
+                                JWT_TOKEN_VALIDATION_FAILED.getMessage(), e);
+                    }
```

Comment:
Shall we capture the respective exceptions in the method level and throw the AuthenticationFailedException from the method instead of capturing Exception e at this level? Any specific reason to do it in this way?

Reply:
Addressed with https://github.com/wso2-extensions/identity-outbound-auth-kakao/pull/15/commits/8d9bd5da9b2a6348daffa96a0e58132d78f39d16